src/utils.rs

```diff
@@ -277,6 +277,83 @@ impl Sample for i32 {
     }
 }
 
+/// An extension-trait to accumulate samples into a frame
+pub(crate) trait FrameAccumulator: Frame {
+    fn scale_add(&mut self, other: &Self, coeff: f32);
+    fn retain_max_samples(&mut self, other: &Self);
+}
+
+impl<F: Frame, S> FrameAccumulator for F
+where
+    S: SampleAccumulator + std::fmt::Debug,
+    F: IndexMut<Target = S>,
+{
+    #[inline(always)]
+    fn scale_add(&mut self, other: &Self, coeff: f32) {
+        for i in 0..Self::CHANNELS {
+            self.index_mut(i).scale_add(*other.index(i), coeff);
+        }
+    }
+
+    fn retain_max_samples(&mut self, other: &Self) {
+        for i in 0..Self::CHANNELS {
+            let this = self.index_mut(i);
+            let other = other.index(i);
+            if *other > *this {
+                *this = *other;
+            }
+        }
+    }
+}
+
+// Required since std::ops::IndexMut seem to not be implemented for arrays
+// making FrameAcc hard to implement for auto-vectorization
```

Comment:
https://github.com/rust-lang/rust/pull/74989

Reply:
Nice! Will add to comment.